### DISCUSSION TITLE
SSG Optimization

### DISCUSSION THREAD

**Discussion by kripainc** [vikejs/vike#1125] [Discussion]
Migrated a large site (100K+ pages - https://recalls.me) from Gatsby to Next and now to VPS. 
Really like the plugin and the freedom it permits to integrate any tech I desire. Thank you for the great Vite plugin.

As the site is large, there are couple of areas where optimization can further improve the site performance:
1. The SSG generated page includes JSON data which may not be necessary. I'm referring to <script id="vite-plugin-ssr_pageContext" type="application/json"> 
Consider the star-wars example which is included in the plugin examples. If you preview source in the browser, you'll see the inclusion of movie data: pageProps":{"movies":[{"title":"A New Hope","release_date":"1977-05-25","id":"1"}, ... ] }  
On RecallsMe site, for a page with 100K+ @id sub-pages, the page ends up with a very large JSON structure (164M). Is there a way to suppress this JSON data? 
2. Incremental Builds - As new recalls are announced by USA Govt, I'll like to include those in my site without rebuilding the whole site. Is there a way to rebuild subset of site and supplementing it to previously built site?

**Comment by brillout** [comment]
> Is there a way to suppress this JSON data?

If you set `passToClient` to an empty array, then the JSON should be almost empty. (Not entirely empty as it includes some internals needed by vite-plugin-ssr.)

> without rebuilding the whole site. Is there a way to rebuild subset of site and supplementing it to previously built site?

Check out Vercel's [ISR](https://vercel.com/docs/incremental-static-regeneration), and [vite-plugin-vercel](https://github.com/magne4000/vite-plugin-vercel) by @magne4000 which supports vite-plugin-ssr.

> Migrated a large site (100K+ pages - https://recalls.me) from Gatsby to Next and now to VPS.

Neat, I very much appreciate such down-to-earth businesses.

> Really like the plugin and the freedom it permits to integrate any tech I desire. Thank you for the great Vite plugin.

I'm glad you like it. Btw. is [sponsoring](https://github.com/sponsors/brillout) something you'd be up to? Could make sense and I'd be up to further improve the performance and DX for your business.

**Reply by kripainc** [reply]
Evidently setting passToClient to an empty array triggers error:  [vite-plugin-ssr][Wrong Usage] pageContext["pageProps"] isn't available on the client-side because "pageProps" is missing in passToClient,
It appears passToClient is tightly bound to functionality of server prerender(). I'll probably need to dig deep into VPS code. 
Any suggestion on where to look?
Further, even though pages/star-wars is SSG, the client side rendering is also triggered (just add a log statement and see it get written to browser console). My understanding was it should only do the React.hydrateRoot() and not do the full fledged render().

**Reply by brillout** [reply]
That's the point: *you* control what get's passed to the client-side. If there is too much data passed to the client-side, then it's an issue on the user land.

> Further, even though pages/star-wars is SSG, the client side rendering is also triggered (just add a log statement and see it get written to browser console). My understanding was it should only do the React.hydrateRoot() and not do the full fledged render().

I don't understand what you mean but it's most likely a user land issue as well.

**Reply by kripainc** [reply]
To clarify, here is a test case:
Update the pages/star-wars/index/index.page.tsx as below with a simple console.log('index.page.tsx - rendering page') and run 'vite preview'.

```

import React from 'react'
import type { Movie } from '../types'

export { Page }

function Page({ movies }: { movies: Movie[] }) {

console.log('index.page.tsx - rendering page');

  return (
    <>
      <h1>Star Wars Movies</h1>


```
﻿For the 'vite build' step, you'll see the log message on your console and the build step works as expected.

However, when you preview it in the browser, the browser console will **also list the above console log statement**. Implying the client side is also rendering.  IE. Running the React's render() function (CSR). 
SSG application should (only do ReactDOM.hydrateRoot() ) not render on client side! 
It contradicts the SSG vs SSR claims at https://vike.dev/pre-rendering
This is not an user land issue.

**Reply by brillout** [reply]
This is expected: `Page` needs to be called for the hydration step to work.
